Use cub::BlockRadixSort to improve medium length sort performance

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #79628

In my testing, replacing the custom bitonic sort with cub's block
level radix sort primitives improves overall sort performance by up to
3x, depending on input length. This also benefits from being a stable
sort, and so we get up to 25x speedup for small stable sorts and
around 2x speedup on the largest supported size.

In testing, the radix sort benefits a lot from having more items per
thread meaning it breaks down a bit at very small sizes. So, for the
32-item sort I've left the bitonic sorting algorithm in place.

Binary size is also reduced in this change, because I have moved the
`descending` branch into the kernel itself which I found not to effect 
performance.  The result is a 1.9 MB decrease in `torch_cuda.so` on
my build for one cuda architecture.

peterbell10 committed:
Use cub::BlockRadixSort to improve medium length sort performance

In my testing, replacing the custom bitonic sort with cub's block
level radix sort primitives improves overall sort performance by up to
3x, depending on input length. This also benefits from being a stable
sort, and so we get up to 25x speedup for small stable sorts and
around 2x speedup on the largest supported size.

In testing, the radix sort benefits a lot from having more items per
thread and so it does break down at very small sizes. So, for the
32-item sort I've left the bitonic sorting algorithm in place.

[ghstack-poisoned]

peterbell10 commented:
Here are my detailed benchmark results for all sizes in `range(0, 4097, 4)`, expressed as speedup relative to the old implementation. All results come from an RTX 2060.

Unstable sort varies from 1-3x speedup
![image](https://user-images.githubusercontent.com/13238737/174122352-b14cd1f7-82c1-4e4d-a40c-8b02bc8140f8.png)

Stable sort varies from 1-25x speedup
![image](https://user-images.githubusercontent.com/13238737/174122469-5a0f7514-cff9-435d-9de5-d8a762e9c36e.png)

peterbell10 committed:
Update on "Use cub::BlockRadixSort to improve medium length sort performance"


In my testing, replacing the custom bitonic sort with cub's block
level radix sort primitives improves overall sort performance by up to
3x, depending on input length. This also benefits from being a stable
sort, and so we get up to 25x speedup for small stable sorts and
around 2x speedup on the largest supported size.

In testing, the radix sort benefits a lot from having more items per
thread meaning it breaks down a bit at very small sizes. So, for the
32-item sort I've left the bitonic sorting algorithm in place.

Binary size is also reduced in this change, because I have moved the
`descending` branch into the kernel itself which I found not to effect 
performance.  The result is a 1.9 MB decrease in `torch_cuda.so` on
my build for one cuda architecture.

[ghstack-poisoned]

peterbell10 committed:
Scale block.y down for small tensors on "Use cub::BlockRadixSort to improve medium length sort performance"


In my testing, replacing the custom bitonic sort with cub's block
level radix sort primitives improves overall sort performance by up to
3x, depending on input length. This also benefits from being a stable
sort, and so we get up to 25x speedup for small stable sorts and
around 2x speedup on the largest supported size.

In testing, the radix sort benefits a lot from having more items per
thread meaning it breaks down a bit at very small sizes. So, for the
32-item sort I've left the bitonic sorting algorithm in place.

Binary size is also reduced in this change, because I have moved the
`descending` branch into the kernel itself which I found not to effect 
performance.  The result is a 1.9 MB decrease in `torch_cuda.so` on
my build for one cuda architecture.

[ghstack-poisoned]

peterbell10 committed:
Update on "Use cub::BlockRadixSort to improve medium length sort performance"


In my testing, replacing the custom bitonic sort with cub's block
level radix sort primitives improves overall sort performance by up to
3x, depending on input length. This also benefits from being a stable
sort, and so we get up to 25x speedup for small stable sorts and
around 2x speedup on the largest supported size.

In testing, the radix sort benefits a lot from having more items per
thread meaning it breaks down a bit at very small sizes. So, for the
32-item sort I've left the bitonic sorting algorithm in place.

Binary size is also reduced in this change, because I have moved the
`descending` branch into the kernel itself which I found not to effect 
performance.  The result is a 1.9 MB decrease in `torch_cuda.so` on
my build for one cuda architecture.

[ghstack-poisoned]

peterbell10 committed:
Don't round batch size up on "Use cub::BlockRadixSort to improve medium length sort performance"


In my testing, replacing the custom bitonic sort with cub's block
level radix sort primitives improves overall sort performance by up to
3x, depending on input length. This also benefits from being a stable
sort, and so we get up to 25x speedup for small stable sorts and
around 2x speedup on the largest supported size.

In testing, the radix sort benefits a lot from having more items per
thread meaning it breaks down a bit at very small sizes. So, for the
32-item sort I've left the bitonic sorting algorithm in place.

Binary size is also reduced in this change, because I have moved the
`descending` branch into the kernel itself which I found not to effect 
performance.  The result is a 1.9 MB decrease in `torch_cuda.so` on
my build for one cuda architecture.

[ghstack-poisoned]

peterbell10 committed:
Rebase on "Use cub::BlockRadixSort to improve medium length sort performance"


In my testing, replacing the custom bitonic sort with cub's block
level radix sort primitives improves overall sort performance by up to
3x, depending on input length. This also benefits from being a stable
sort, and so we get up to 25x speedup for small stable sorts and
around 2x speedup on the largest supported size.

In testing, the radix sort benefits a lot from having more items per
thread meaning it breaks down a bit at very small sizes. So, for the
32-item sort I've left the bitonic sorting algorithm in place.

Binary size is also reduced in this change, because I have moved the
`descending` branch into the kernel itself which I found not to effect 
performance.  The result is a 1.9 MB decrease in `torch_cuda.so` on
my build for one cuda architecture.

[ghstack-poisoned]

ngimel commented on aten/src/ATen/native/cuda/SortUtils.cuh:
```diff
@@ -151,4 +153,101 @@ bitonicSortKVInPlace(at::cuda::detail::TensorInfo<K, IndexType> keys,
+          typename K, typename V, typename IndexType>
+C10_LAUNCH_BOUNDS_1(block_size)
+__global__ void
+radixSortKVInPlace(at::cuda::detail::TensorInfo<K, IndexType> keys,
```
is this better than DeviceSegmentedRadix sort that's used for some configurations already? It seems like it should be pretty similar.

peterbell10 replied:
Yes, it is significantly better. `DeviceSegmentedRadixSort` is used by `launch_stable_sort_kernel` and if you look at the stable sort speedup graph you can see `radixSortKVInPlace` is at worst 1.5x faster and at best 25x faster, depending length of the dimension being sorted.

ngimel replied:
Hm, interesting, seems like cub should have gotten it right

ngimel commented:
@pytorchbot merge

pytorchmergebot committed:
Update on "Use cub::BlockRadixSort to improve medium length sort performance"


In my testing, replacing the custom bitonic sort with cub's block
level radix sort primitives improves overall sort performance by up to
3x, depending on input length. This also benefits from being a stable
sort, and so we get up to 25x speedup for small stable sorts and
around 2x speedup on the largest supported size.

In testing, the radix sort benefits a lot from having more items per
thread meaning it breaks down a bit at very small sizes. So, for the
32-item sort I've left the bitonic sorting algorithm in place.

Binary size is also reduced in this change, because I have moved the
`descending` branch into the kernel itself which I found not to effect 
performance.  The result is a 1.9 MB decrease in `torch_cuda.so` on
my build for one cuda architecture.

[ghstack-poisoned]

peterbell10 committed:
Update on "Use cub::BlockRadixSort to improve medium length sort performance"


In my testing, replacing the custom bitonic sort with cub's block
level radix sort primitives improves overall sort performance by up to
3x, depending on input length. This also benefits from being a stable
sort, and so we get up to 25x speedup for small stable sorts and
around 2x speedup on the largest supported size.

In testing, the radix sort benefits a lot from having more items per
thread meaning it breaks down a bit at very small sizes. So, for the
32-item sort I've left the bitonic sorting algorithm in place.

Binary size is also reduced in this change, because I have moved the
`descending` branch into the kernel itself which I found not to effect 
performance.  The result is a 1.9 MB decrease in `torch_cuda.so` on
my build for one cuda architecture.

[ghstack-poisoned]

peterbell10 commented:
@pytorchbot merge -g

peterbell10 committed:
Rebase on "Use cub::BlockRadixSort to improve medium length sort performance"


In my testing, replacing the custom bitonic sort with cub's block
level radix sort primitives improves overall sort performance by up to
3x, depending on input length. This also benefits from being a stable
sort, and so we get up to 25x speedup for small stable sorts and
around 2x speedup on the largest supported size.

In testing, the radix sort benefits a lot from having more items per
thread meaning it breaks down a bit at very small sizes. So, for the
32-item sort I've left the bitonic sorting algorithm in place.

Binary size is also reduced in this change, because I have moved the
`descending` branch into the kernel itself which I found not to effect 
performance.  The result is a 1.9 MB decrease in `torch_cuda.so` on
my build for one cuda architecture.

[ghstack-poisoned]

peterbell10 commented:
@pytorchbot merge -g

peterbell10 committed:
Attempt to fix rocm build on "Use cub::BlockRadixSort to improve medium length sort performance"


In my testing, replacing the custom bitonic sort with cub's block
level radix sort primitives improves overall sort performance by up to
3x, depending on input length. This also benefits from being a stable
sort, and so we get up to 25x speedup for small stable sorts and
around 2x speedup on the largest supported size.

In testing, the radix sort benefits a lot from having more items per
thread meaning it breaks down a bit at very small sizes. So, for the
32-item sort I've left the bitonic sorting algorithm in place.

Binary size is also reduced in this change, because I have moved the
`descending` branch into the kernel itself which I found not to effect 
performance.  The result is a 1.9 MB decrease in `torch_cuda.so` on
my build for one cuda architecture.

[ghstack-poisoned]

peterbell10 commented:
@pytorchbot merge
